### DISCUSSION TITLE
jdbc integration

### DISCUSSION THREAD

**Discussion by devkral** [sqlalchemy/sqlalchemy#11684] [Blue Sky]
After weeks of hard work I managed to build a jdbc integration for sqlalchemy.

It is based on the dbapi2 module of Jpype. It is in the new databasez release of dymmond.
Using jdbc still requires using the project itself because I do dsn rewrites so sqlalchemy receives a sanitized dsn, I lazily inject the dialects and I start the jvm in my code overwrites. Most things I think are fixable with better sqlalchemy knowledge but despite a really nice in code documentation I found no tutorial in how to write a dialect.

But maybe we get patches and guidance in how to improve id.

**Comment by devkral** [comment]
Not to mention: the project contains also a generic dbapi2 dialect (I know you can inject dbapi2 modules in dialects). The sweet part is:
it can wrap a non async dbapi and make it async.

But again, improvements are possible

**Comment by CaselIT** [comment]
Hi,

Have you looked at https://github.com/sqlalchemy/sqlalchemy/blob/main/README.dialects.rst ?

**Reply by devkral** [reply]
this is a stub. It doesn't explain what methods to implement and how to derive from the DefaultDialect.

The inline code documentation is awesome but trial and error.

**Reply by devkral** [reply]
for example:

has_table has no default implementation. I had to figure out an awkward way with sqlalchemy_utils.

**Reply by CaselIT** [reply]
by definition has_table depends on the db you target. there isn't a default implementation since there avery db does things differently.

**Reply by devkral** [reply]
There is an ugly way to check it: query the table and see if it fails.

**Reply by devkral** [reply]
I am aware of the consequences. You may need to reset a transaction and performance is bad.

**Reply by CaselIT** [reply]
I disagree with it being ugly. You are free to implement that way though.

**Reply by zzzeek** [reply]
why not look at any of the other [many dozens of dialect implementations](https://docs.sqlalchemy.org/en/20/dialects/index.html) to see what some common practices are for this method?

**Reply by devkral** [reply]
I did so. But sqlalchemy evolves fast so the dialects doesn't keep pace and define many unnecessary methods.

**Reply by devkral** [reply]
A clean list of methods and attributes in use would be helpful. As well as a async dbapi2 translation layer but this is a different story.

**Reply by zzzeek** [reply]
the clean list of methods and attributes is here:  https://docs.sqlalchemy.org/en/20/core/internals.html#sqlalchemy.engine.Dialect 

are there required attributes or methods missing from there?  maybe, if so, those are bugs to be reported here

**Reply by CaselIT** [reply]
> I did so. But sqlalchemy evolves fast so the dialects doesn't keep pace and define many unnecessary methods.

I'm not aware of any unnecessary method in the internal dialect implementations.

**Reply by devkral** [reply]
thank you! Missed this part in the page.

But I really would wish for some simpler documentation like django has them.

**Reply by zzzeek** [reply]
if you want simpler docs then use django :)   this is what we have and for a motivated developer they are more than enough

**Reply by CaselIT** [reply]
If you find ways to improve them, even only part of a method or two, docs PR are always welcome
